mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -0,0 +1,727 @@
+use mqtt3::proto::Publication;
+use serde::{Deserialize, Serialize};
+use tracing::error;
+
+use crate::persist::{
+    waking_state::ring_buffer::{
+        block::{
+            calculate_hash, serialized_block_size, validate, BlockHeaderWithHash, Data, BLOCK_HINT,
+        },
+        error::RingBufferError,
+        flush::{FlushOptions, FlushState},
+        serialize::{binary_deserialize, binary_serialize, binary_serialize_size},
+    },
+    Key, StorageError,
+};
+
+#[allow(dead_code)]
+pub type StorageResult<T> = Result<T, StorageError>;
+
+/// Convenience struct for tracking read and write pointers into the file.
+#[allow(dead_code)]
+#[derive(Debug, Default, Deserialize, Serialize)]
+pub(crate) struct FilePointers {
+    write: usize,
+    read: usize,
+}
+
+/// Imagine there are three pointers:
+/// W - A write pointer
+/// R - A read pointer
+/// D - A delete pointer
+///
+/// With a file buffer similar to below:
+/// +---+---+---+---+---+---+---+---+---+
+/// | A | B | C | D | E | F | G | H | I |
+/// +---+---+---+---+---+---+---+---+---+
+///
+/// It is possible that W could wrap around and cause us to lose data due to
+/// variable sized data being written. If W + data size > R and should write
+/// precede read, the data could become corrupted.
+///
+/// Either this is prevented by scanning the file for the next block and
+/// checking if the write will fit. Or, W + data must not pass both D and R.
+/// - Not passing D helps with guaranteeing all data read was handled.
+/// - Not passing R helps avoid corruption.
+#[allow(dead_code)]
+pub(crate) struct RingBuffer<'a> {
+    flush_options: FlushOptions,
+    flush_state: FlushState,
+    mmap: &'a mut MmapMut,
+    max_file_size: usize,
+    order: usize,
+    pointers: FilePointers,
+    waker: Option<Waker>,
+}
+
+#[allow(dead_code)]
+impl<'a> RingBuffer<'a> {
+    fn new(mmap: &'a mut MmapMut, max_file_size: usize, flush_options: FlushOptions) -> Self {
+        let file_pointers =
+            find_pointers_post_crash(&mmap, max_file_size).expect("Failed to init file pointers");
+        Self {
+            flush_options,
+            flush_state: FlushState::new(),
+            mmap,
+            max_file_size,
+            order: usize::default(),
+            pointers: file_pointers,
+            waker: None,
+        }
+    }
+
+    fn insert(&mut self, publication: &Publication) -> StorageResult<Key> {
+        let timer = Instant::now();
+        let data = binary_serialize(publication)?;
+        let data = Data::new(&data);
+        let data_size = binary_serialize_size(&data)?;
+        let data_hash = calculate_hash(&data);
+
+        let write_index = self.pointers.write;
+        let read_index = self.pointers.read;
+        let order = self.order;
+        let key = write_index;
+
+        let block_header =
+            BlockHeaderWithHash::new(data_hash, data_size, order, read_index, write_index);
+        let block_size = serialized_block_size()?;
```

Comment:
can this be static?

Reply:
I want it to be :) I need a way to get it to be static through bincode though... maybe this is something lazy_static can do? Haven't used it before, so if you know of a way to do this correctly, let me know!

Reply:
you can add this somewhere at the top of the rign_buffer/mod.rs
```rust
lazy_static! {
    static ref BLOCK_SIZE: usize =
        serialized_block_size().expect("failed to calculate ring buffer block size");
}
```
and use it like this:
```rust
let block_size = *BLOCK_SIZE;
```